### PR TITLE
Eliminate lint warnings

### PR DESCRIPTION
**Pull request by roienatan**
resolves: https://github.com/daostack/alchemy/issues/2016
resolves: https://github.com/daostack/alchemy/issues/1867

**IMPORTENT NOTE: for any future development please check with `rpm run lint` before pushing your code. Use `eslint-disable` as a last resort.**

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -28,13 +28,15 @@
         "max-len": "off",
         "react/display-name" : "off",
         "sort-imports": "off",
+        // Most of the warnings arises from inline arrow functions in Formik.
+        // As long as we use Formik looks like we can't get rid of this warning.
+        "react/jsx-no-bind": "off",
```
Comment (dkent600) [inline .eslintrc:33]: I would prefer not to turn this rule off.  It is a good rule, no?  For the formik warnings we can do a one-line disable with a eslint comment.

File: src/actions/profilesActions.ts [hunk 2]
```diff
@@ -162,7 +163,7 @@ export function updateProfile(accountAddress: string, name: string, description:
 export type FollowItemAction = IAsyncAction<"FOLLOW_ITEM", { accountAddress: string }, { type: FollowType; id: string; isFollowing: boolean}>;
 
 export function toggleFollow(accountAddress: string, type: FollowType, id: string) {
-  return async (dispatch: any, _getState: any) => {
+  return async (dispatch: any, _getState: any): Promise<boolean> => {
```
Comment (dkent600) [inline src/actions/profilesActions.ts:166]: All of these function params should be `(dispatch: Redux.Dispatch<any, any>, _getState: () => IRootState)`

File: src/arc.ts [hunk 3]
```diff
@@ -421,6 +421,7 @@ async function getCurrentAccountFromProvider(): Promise<Address | null> {
  * Clear caches as every case where we're manually logging out
  * implies that the cache should be cleared
  */
+//eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```
Comment (dkent600) [inline src/arc.ts:424]: why is this needed?  I suspect it is because you didn't figure out a better alternative to `any`. I think there likely _is_ a better way and would prefer that we find it.

File: src/components/Buidlhub/BuidlhubClient.ts [hunk 4]
```diff
@@ -37,7 +37,8 @@ export default class BuidlhubClient implements IBuidlhubClient {
     }
   }
 
-  async _post(body: any) {
+  //eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```
Comment (dkent600) [inline src/components/Buidlhub/BuidlhubClient.ts:40]: why is this needed? I suspect it is because you didn't figure out a better alternative to any. I think there likely is a better way and would prefer that we find it.

File: src/components/Plugin/ContributionRewardExtRewarders/Competition/Details.tsx [hunk 5]
```diff
@@ -443,6 +443,7 @@ export default withSubscription({
     ${CompetitionSuggestion.fragments.CompetitionSuggestionFields}
     `;
 
+    // eslint-disable-next-line no-console
```
Comment (dkent600) [inline src/components/Plugin/ContributionRewardExtRewarders/Competition/Details.tsx:446]: lets delete this console.log?

File: src/components/Shared/withSubscription.tsx [hunk 6]
```diff
@@ -68,6 +68,7 @@ interface IWithSubscriptionOptions<Props extends ISubscriptionProps<ObservableTy
  * required by the wrappedComponent and passes along the data observed as it comes in.
  * @oarams options See `IWithSubscriptionOptions`
  */
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```
Comment (dkent600) [inline src/components/Shared/withSubscription.tsx:71]: why is this needed?

File: src/lib/analytics.ts [hunk 7]
```diff
@@ -1,3 +1,4 @@
+/* eslint @typescript-eslint/explicit-module-boundary-types: 0 */
```
Comment (dkent600) [inline src/lib/analytics.ts:1]: Why is this needed?

File: src/reducers/profilesReducer.ts [hunk 8]
```diff
@@ -48,7 +48,8 @@ export function newProfile(ethereumAccountAddress: string): IProfileState {
 
 const initialState: IProfilesState = { threeBox: null, threeBoxSpace: null };
 
-const profilesReducer = (state = initialState, action: any) => {
+//eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```
Comment (dkent600) [inline src/reducers/profilesReducer.ts:51]: why is this needed? I suspect it is because you didn't figure out a better alternative to any. I think there likely is a better way and would prefer that we find it.

File: src/reducers/uiReducer.ts [hunk 9]
```diff
@@ -31,7 +31,8 @@ const initialState: IUIState = {
   trainingTooltipsShowAll: false,
 };
 
-const uiReducer = (state = initialState, action: any) => {
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```
Comment (dkent600) [inline src/reducers/uiReducer.ts:34]: why is this needed? I suspect it is because you didn't figure out a better alternative to any. I think there likely is a better way and would prefer that we find it.

File: src/reducers/web3Reducer.ts [hunk 10]
```diff
@@ -15,7 +15,8 @@ const initialState: IWeb3State = {
   networkId: 0, // unknown network
 };
 
-const web3Reducer = (state = initialState, action: any) => {
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```
Comment (dkent600) [inline src/reducers/web3Reducer.ts:18]: why is this needed? I suspect it is because you didn't figure out a better alternative to any. I think there likely is a better way and would prefer that we find it.

File: test/integration/utils.ts [hunk 11]
```diff
@@ -57,10 +57,11 @@ export function getTestAddresses(version: string = LATEST_ARC_VERSION): ITestAdd
   return require("@daostack/test-env-experimental/daos.json").demo[version];
 }
 
-export async function hideTrainingTooltips() {
+export async function hideTrainingTooltips(): Promise<void> {
   localStorage.setItem("trainingTooltipsEnabled", "false");
 }
 
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```
Comment (dkent600) [inline test/integration/utils.ts:64]: why is this needed? I suspect it is because you didn't figure out a better alternative to any. I think there likely is a better way and would prefer that we find it.

File: src/actions/profilesActions.ts [hunk 12]
```diff
@@ -1,3 +1,4 @@
+/* eslint @typescript-eslint/explicit-module-boundary-types: 0 */
```
Comment (dkent600) [inline src/actions/profilesActions.ts:1]: why is this needed?

File: src/components/Errors/Error404.tsx [hunk 13]
```diff
@@ -5,14 +5,15 @@ import * as css from "./Errors.scss";
 
 export default class Error404 extends React.PureComponent {
 
+  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
```
Comment (dkent600) [inline src/components/Errors/Error404.tsx:8]: why is this needed? I suspect it is because you didn't figure out a better alternative to any. I think there likely is a better way and would prefer that we find it.

File: src/components/Plugin/ContributionRewardExtRewarders/Competition/CreateProposal.tsx [hunk 14]
```diff
@@ -225,7 +225,7 @@ class CreateProposal extends React.Component<IProps, IStateProps> {
         <Formik
           // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
           initialValues={this.initialFormValues}
-          // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Plugin/ContributionRewardExtRewarders/Competition/CreateProposal.tsx:228]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Plugin/ContributionRewardExtRewarders/Competition/CreateProposal.tsx [hunk 15]
```diff
@@ -340,7 +340,6 @@ class CreateProposal extends React.Component<IProps, IStateProps> {
             return errors;
           }}
           onSubmit={this.handleSubmit}
-          // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Plugin/ContributionRewardExtRewarders/Competition/CreateProposal.tsx:343]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Plugin/ContributionRewardExtRewarders/Competition/CreateSubmission.tsx [hunk 16]
```diff
@@ -92,7 +92,7 @@ class CreateSubmission extends React.Component<IProps, IStateProps> {
         <Formik
           // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
           initialValues={this.initialFormValues}
-          // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Plugin/ContributionRewardExtRewarders/Competition/CreateSubmission.tsx:95]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Plugin/ContributionRewardExtRewarders/Competition/CreateSubmission.tsx [hunk 17]
```diff
@@ -120,7 +120,7 @@ class CreateSubmission extends React.Component<IProps, IStateProps> {
             return errors;
           }}
           onSubmit={this.handleSubmit}
-          // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Plugin/ContributionRewardExtRewarders/Competition/CreateSubmission.tsx:123]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Plugin/ReputationFromToken.tsx [hunk 18]
```diff
@@ -284,7 +284,6 @@ class ReputationFromToken extends React.Component<IProps, IState> {
               useTxSenderService: false,
             } as IFormValues}
 
-            // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Plugin/ReputationFromToken.tsx:287]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Plugin/ReputationFromToken.tsx [hunk 19]
```diff
@@ -305,7 +304,6 @@ class ReputationFromToken extends React.Component<IProps, IState> {
 
             onSubmit={this.handleSubmit}
 
-            // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Plugin/ReputationFromToken.tsx:308]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreateContributionRewardProposal.tsx [hunk 20]
```diff
@@ -179,7 +179,7 @@ class CreateContributionReward extends React.Component<IProps, IStateProps> {
         <Formik
           // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
           initialValues={this.initialFormValues}
-          // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreateContributionRewardProposal.tsx:182]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreateContributionRewardProposal.tsx [hunk 21]
```diff
@@ -221,11 +221,10 @@ class CreateContributionReward extends React.Component<IProps, IStateProps> {
             return errors;
           }}
           onSubmit={this.handleSubmit}
-          // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreateContributionRewardProposal.tsx:224]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreateKnownGenericPluginProposal.tsx [hunk 22]
```diff
@@ -188,7 +188,6 @@ class CreateKnownPluginProposal extends React.Component<IProps, IState> {
         </div>;
       default:
         if (field.type.includes("[]")) {
-          // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreateKnownGenericPluginProposal.tsx:191]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreateKnownGenericPluginProposal.tsx [hunk 23]
```diff
@@ -309,7 +308,6 @@ class CreateKnownPluginProposal extends React.Component<IProps, IState> {
         <div className={css.contentWrapper}>
           <Formik
             initialValues={this.initialFormValues}
-            // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreateKnownGenericPluginProposal.tsx:312]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreateKnownGenericPluginProposal.tsx [hunk 24]
```diff
@@ -381,7 +379,6 @@ class CreateKnownPluginProposal extends React.Component<IProps, IState> {
               return errors;
             }}
             onSubmit={this.handleSubmit}
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreateKnownGenericPluginProposal.tsx:381]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreatePluginManagerProposal.tsx [hunk 25]
```diff
@@ -502,7 +502,6 @@ class CreatePluginManagerProposal extends React.Component<IProps, IState> {
           <Formik
             // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
             initialValues={this.initialFormValues}
-            // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreatePluginManagerProposal.tsx:505]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreatePluginManagerProposal.tsx [hunk 26]
```diff
@@ -535,7 +534,6 @@ class CreatePluginManagerProposal extends React.Component<IProps, IState> {
               return errors;
             }}
             onSubmit={this.handleSubmit}
-            // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreatePluginManagerProposal.tsx:538]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreatePluginRegistrarProposal.tsx [hunk 27]
```diff
@@ -267,7 +267,6 @@ class CreatePluginRegistrarProposal extends React.Component<IProps, IState> {
             <Formik
               // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
               initialValues={this.initialFormValues}
-              // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreatePluginRegistrarProposal.tsx:270]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreatePluginRegistrarProposal.tsx [hunk 28]
```diff
@@ -306,7 +305,6 @@ class CreatePluginRegistrarProposal extends React.Component<IProps, IState> {
                 return errors;
               }}
               onSubmit={this.handleSubmit}
-              // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreatePluginRegistrarProposal.tsx:309]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreateUnknownGenericPluginProposal.tsx [hunk 29]
```diff
@@ -107,7 +107,6 @@ class CreateGenericPlugin extends React.Component<IProps, IStateProps> {
         <Formik
           // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
           initialValues={this.initialFormValues}
-          // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreateUnknownGenericPluginProposal.tsx:110]: See earlier comment, would it not be better to keep this rule enabled?

File: src/components/Proposal/Create/PluginForms/CreateUnknownGenericPluginProposal.tsx [hunk 30]
```diff
@@ -152,7 +151,6 @@ class CreateGenericPlugin extends React.Component<IProps, IStateProps> {
             return errors;
           }}
           onSubmit={this.handleSubmit}
-          // eslint-disable-next-line react/jsx-no-bind
```
Comment (dkent600) [inline src/components/Proposal/Create/PluginForms/CreateUnknownGenericPluginProposal.tsx:155]: See earlier comment, would it not be better to keep this rule enabled?

**Comment by dkent600** [comment]
Regarding the comment in the description about running lint:  Travis will fail if `npm run lint` fails.  If this is not the case then we need to fix something.

It won't fail on warnings.  But warnings are warnings and if we consider warnings to be errors then we should change the lint rules.

**Comment by roienatan** [comment]
> Regarding the comment in the description about running lint: Travis will fail if `npm run lint` fails. If this is not the case then we need to fix something.
> 
> It won't fail on warnings. But warnings are warnings and if we consider warnings to be errors then we should change the lint rules.

Sure it fails on errors and gives warnings but I say in order to prevent a case that we need to modify more than 40 files to eliminate the warnings, it makes more sense that before pushing any piece of code, each should check his changes.

File: .eslintrc (hunk 1 above)
Comment (roienatan) [reply .eslintrc:33]: Agree, this is a good rule. I was not sure if to disable it globally (since most of them related to Formik) or disable specifically. I didn't want to add many lines in the middle of the code. What do you say about disabling for specific files? It's a one line in the top of the file.

File: src/actions/profilesActions.ts (hunk 2 above)
Comment (roienatan) [reply src/actions/profilesActions.ts:166]: Thanks!

File: src/components/Plugin/ContributionRewardExtRewarders/Competition/Details.tsx (hunk 5 above)
Comment (roienatan) [reply src/components/Plugin/ContributionRewardExtRewarders/Competition/Details.tsx:446]: Meaning there is no need in the `cacheQuery` from graphQL? this `console.log` is the only one who uses this.

File: .eslintrc (hunk 1 above)
Comment (dkent600) [reply .eslintrc:33]: I think it best to disable it line-by line, as we were doing.  Thus it is clear in the code and we don't risk disabling for cases that we should fix.

File: src/components/Plugin/ContributionRewardExtRewarders/Competition/Details.tsx (hunk 5 above)
Comment (dkent600) [reply src/components/Plugin/ContributionRewardExtRewarders/Competition/Details.tsx:446]: Oh yeah, this reminds me that I've looked at this before.  @jellegerbrandy added the `console.log` and commented out the code that actually executes the query.  I don't know why.  Jelle?

File: src/reducers/profilesReducer.ts (hunk 8 above)
Comment (roienatan) [reply src/reducers/profilesReducer.ts:51]: As we spoke, this one I leave for you as you're working on this file and you've a solution for this.

**Comment by orenyodfat** [comment]
please update travis to report/fail also on warnings. - remove quiet.

File: test/integration/utils.ts (hunk 11 above)
Comment (roienatan) [reply test/integration/utils.ts:64]: Looks like there is some problem with `webdriverio` and `typescript`, it keeps saying that `setValue` exist only on type `any`.

File: src/components/Shared/withSubscription.tsx [hunk 31]
```diff
@@ -68,7 +68,8 @@ interface IWithSubscriptionOptions<Props extends ISubscriptionProps<ObservableTy
  * required by the wrappedComponent and passes along the data observed as it comes in.
  * @oarams options See `IWithSubscriptionOptions`
  */
-const withSubscription = <Props extends ISubscriptionProps<ObservableType>, ObservableType extends Record<string, any>>(options: IWithSubscriptionOptions<Props, ObservableType>) => {
+
+const withSubscription = <Props extends ISubscriptionProps<ObservableType>, ObservableType extends Record<string, any>>(options: IWithSubscriptionOptions<Props, ObservableType>): any => {
```
Comment (dkent600) [inline src/components/Shared/withSubscription.tsx:72]: This returns:

`new (...args: any) => React.Component<OnlyWrappedComponentProps<Props>, IObservableState<ObservableType>>`

Comment (roienatan) [reply src/components/Shared/withSubscription.tsx]: When trying this return value I get errors in `CreateProposalPage.tsx`
